modules/genericAnalyticsAdapter.js

```diff
@@ -0,0 +1,155 @@
+import AnalyticsAdapter from '../libraries/analyticsAdapter/AnalyticsAdapter.js';
+import {prefixLog, isPlainObject} from '../src/utils.js';
+import * as CONSTANTS from '../src/constants.json';
+import adapterManager from '../src/adapterManager.js';
+import {ajaxBuilder} from '../src/ajax.js';
+import {VENDORLESS_GVLID} from '../src/consentHandler.js';
+
+const DEFAULTS = {
+  batchSize: 1,
+  batchDelay: 100,
+  method: 'POST'
+}
+
+const TYPES = {
+  handler: 'function',
+  batchSize: 'number',
+  batchDelay: 'number',
+}
+
+const MAX_CALL_DEPTH = 20;
+
+export function GenericAnalytics() {
+  const parent = AnalyticsAdapter({analyticsType: 'endpoint'});
+  const {logError, logWarn} = prefixLog('Generic analytics:');
+  let batch = [];
+  let callDepth = 0;
+  let options, handler, timer, translate;
+
+  function optionsAreValid(options) {
+    if (!options.url && !options.handler) {
+      logError('options must specify either `url` or `handler`')
+      return false;
+    }
+    if (options.hasOwnProperty('method') && !['GET', 'POST'].includes(options.method)) {
+      logError('options.method must be GET or POST');
+      return false;
+    }
+    for (const [field, type] of Object.entries(TYPES)) {
+      // eslint-disable-next-line valid-typeof
+      if (options.hasOwnProperty(field) && typeof options[field] !== type) {
+        logError(`options.${field} must be a ${type}`);
+        return false;
+      }
+    }
+    if (options.hasOwnProperty('events')) {
+      if (!isPlainObject(options.events)) {
+        logError('options.events must be an object');
+        return false;
+      }
+      for (const [event, handler] of Object.entries(options.events)) {
+        if (!CONSTANTS.EVENTS.hasOwnProperty(event)) {
+          logWarn(`options.events.${event} does not match any known Prebid event`);
+          if (typeof handler !== 'function') {
+            logError(`options.events.${event} must be a function`);
+            return false;
+          }
+        }
+      }
+    }
+    return true;
+  }
+
+  function processBatch() {
+    const currentBatch = batch;
+    batch = [];
+    callDepth++;
+    try {
+      // the pub-provided handler may inadvertently cause an infinite chain of events;
+      // even just logging an exception from it may cause an AUCTION_DEBUG event, that
+      // gets back to the handler, that throws another exception etc.
+      // to avoid the issue, put a cap on recursion
```

Comment:
these infinite loops have occurred before, eg https://github.com/prebid/Prebid.js/pull/7805/files ; should we more generally defend against them?

Reply:
There is a more general prevention mechanism in https://github.com/prebid/Prebid.js/pull/9113. That also allows the pub to set which events should be forwarded to any analytics adapter. However, adapters can choose to ignore some or circumvent the default event tracking system; so in general, the only way to find out which events are tracked is reading code. For this particular adapter though control is entirely on the publisher.